### DISCUSSION TITLE
Please help writing objects

### DISCUSSION THREAD

**Discussion by chlorid** [ChristianTremblay/BAC0#509] [Q&A] [answered]
Hey there,

first of all thanks for the software I apreciate the work.
I was lucky at work and got a huge box full of Equipment of a company called Merkur. It's all 433 MHZ based radio stuff (temperature sensors of different kinds, MBus adaptors, repeaters and actors). It is all proprietarty and encrypted, so best is to use the base tha was shipped with the stuff. Ah and the company is bancrupt, so zero support to be expected. It is a BacNet base.
Using yabe, I cann subscribe and write data points of the actor and temperature sensors.
Using BAC0, I can read the actor's values, as well as the temperature sensors without problem. 
This is my code:

```python

async def bac(conf):
    print(BAC0.infos.__version__) # 2024.9.20
    async with BAC0.connect(conf["BACNET"]["ownIP"]) as bacnet:
        while True:
            try:
                result = await bacnet.readMultiple(conf["BACNET"]["address"],request_dict=conf["BACNET"])
                for objectName, objectData in conf["BACNET"]["objectMetadata"].items():
                    print("objectName:", objectName, "objectData:", objectData)

                time.sleep(conf["BACNET"]["pollInterval"])
            except Exception as e:
                print(f"can't read {attempt}/5: {e}")
                time.sleep(conf["BACNET"]["pollInterval"])

```

The problem I'm having now is that I cannot use writeMultiple as the function just doesn't seem to exist (anymore?).

I tried then reading and writing single data points but I have the problem that I can also read that way, but not write. I also experience the issue that at some point the connection seems to not have been closed properly. That leads to BAC0 being unable to receive packages from the BacNet device. The only fix I found is restarting the machine, which is super annoying. Weirdly I dont have this problem using above code that doesn't even contain a disconnect().

Maybe that's a particular issue of my kernel, though. 
I'm using a Macbook Pro M1 with Macos Sonoma 14.6.1

This is the code:
```python

async def bacTest(conf):

    bacnet = BAC0.connect(conf["BACNET"]["ownIP"])
    mycontroller = await BAC0.device('192.168.177.29', 1010393, bacnet)
    BAC0.log_level(stdout='debug')
    
    print("request1**********************+")
    res = await mycontroller["S3_Value"].read_property("presentValue")
    # some other tries
    #res = await bacnet.read("192.168.177.29 binaryValue 1024 presentValue")
    #res = await bacnet.read("192.168.177.29 analogValue 1 presentValue")
    print("res1:", res)
    print("write**********************+")
    await mycontroller["S3_Value"].write(99.9)
    # bacnet.write(conf["BACNET"]["address"] + 'binaryValue 1024 presentValue active - 8')
    print("request2**********************+")
    #res = await bacnet.read("192.168.177.29 binaryValue 1024 presentValue")
    res = await mycontroller["S3_Value"].read_property("presentValue")
    print("res2:", res)
    print("**********************+")
    bacnet.disconnect()

```

I'm not super sure if I did anything right or if I have a simple mistake somewhere. Anyway I have the feeling that not all chapters of the documentation have been updated after switching to asyncio. Is that possible? I'm using asyncio fo the first time so it's also possible that I'm doing somethign wrong.

Any help on how I can write data points and avoid my program closing the socket properly would be greatly apreciated.
Maybe some one has a minimal working example on how to read and write a data point?

Best, chlorid

P.S.:
what I also noted:

I tried the subscribe_cov feature with a callback function but the entire callback thing does not seem to be implemeted. At least the bacnet.cov() function does not even have the callback parameter as described in the docs. Is the mechanism deprecated since the program shifted to asyncio?

**Comment by ChristianTremblay** [comment, marked as answer]
Not all controllers support writeMultiple. 

```
print("write**********************+")
    await mycontroller["S3_Value"]._write(99.9) # add underscore here... explanation below
```
Because I'm learning asyncio... and my attempt was to have bacnet.write kinda synchronous... because at first I wanted to reduce the number of "await" required to use BAC0....
bacnet._write is the async part.

And bacnet.write do not create the task everytime.... or rarely... some work needed

Also, I see you used time.sleep... as its async, you should use
```
await asyncio.sleep(1)
```

Or best use a polling task... but let's keep it simple for now

**Reply by chlorid** [reply]
Hi @ChristianTremblay ,
sorry for the late answer, I was on workation and therefor my proivate projects came to a hold.

I got everything working in the meantime, thank you for the suggestions.
It was a lot of trial and error, but finally it runs. Here is my loop:
```

async def bac(conf):
    mqttObj = conf["MQTT"]["mqttObj"]
    print(BAC0.infos.__version__) # 2024.9.20
    async with BAC0.connect(conf["BACNET"]["ownIP"]) as bacnet:
        while True:
            # there is a routing to the external IP to the internal one
            try:
                msgs = []
                topics = []
                result = await bacnet.readMultiple(conf["BACNET"]["address"],request_dict=conf["BACNET"])
                for objectName, objectData in conf["BACNET"]["objectMetadata"].items():
                    print("objectName:", objectName, "objectData:", objectData)
                    if objectName in result:
                        print("objname in result", float(result[objectName][0][1]))
                        topics.append(objectData["sendTopic"])

                        msgs.append(str(result[objectName][0][1]))
                        if "dbObject" in conf["DB"]:
                            conf["DB"]["dbObject"].send_sensor_data_to_influxdbJson({}, objectData["dbMeasurementName"], {objectData["dbValueName"] : float(result[objectName][0][1])})

                print("Result", json.dumps(result, indent=2))
                print("MSGs", str(msgs))
                print("TOPICS", str(topics))
                mqttObj.sendMqttMessage(msgs, topics)
                
                await asyncio.sleep(conf["BACNET"]["pollInterval"])
                #time.sleep(conf["BACNET"]["pollInterval"])
            except Exception as e:
                print(f"can't read {attempt}/5: {e}")
                await asyncio.sleep(conf["BACNET"]["pollInterval"])
                time.sleep(conf["BACNET"]["pollInterval"])
```

the request dict looks like this:
```

 "objects": {
            "analogValue:0": ["objectName", "presentValue"],
            "analogValue:1": ["objectName", "presentValue"],
            "analogValue:2": ["objectName", "presentValue"]
        }
```

I hope this can be of help for someone wanting to use the software.

Best, chlorid
